### PR TITLE
carbon vortex - handle edge case

### PR DESCRIPTION
**Pull request by ivanzhelyazkov**
* Handle edge case when user sends unnecessary native tokens if the token traded is not native
* Add tests

### CODE REVIEW

**Review by yudilevi — APPROVED** [review]

File: contracts/vortex/CarbonVortex.sol [hunk 1]
```diff
@@ -675,6 +679,10 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
         if (sourceAmount == 0) {
             revert InvalidTrade();
         }
+        // revert if unnecessary native token is received
+        if (_finalTargetToken != NATIVE_TOKEN && msg.value > 0) {
+            revert UnnecessaryNativeTokenReceived();
+        }
```
Comment (barakman) [inline contracts/vortex/CarbonVortex.sol]: You can put this in the `else` a few lines below, i.e.:
```
        if (_finalTargetToken == NATIVE_TOKEN) {
            if (msg.value < sourceAmount) {
                revert InsufficientNativeTokenSent();
            }
            payable(_transferAddress).sendValue(sourceAmount);
        } else {
            if (msg.value > 0) {
                revert UnnecessaryNativeTokenReceived();
            }
            _finalTargetToken.safeTransferFrom(msg.sender, _transferAddress, sourceAmount);
        }
```

Comment (ivanzhelyazkov) [reply contracts/vortex/CarbonVortex.sol]: fixed

File: contracts/vortex/CarbonVortex.sol [hunk 2]
```diff
@@ -631,6 +631,10 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
         if (sourceAmount == 0) {
             revert InvalidTrade();
         }
+        // revert if unnecessary native token is received
+        if (_targetToken != NATIVE_TOKEN && msg.value > 0) {
+            revert UnnecessaryNativeTokenReceived();
+        }
```
Comment (barakman) [inline contracts/vortex/CarbonVortex.sol:637]: Can turn this into an if/else construct, similar to the one in the other function, i.e.:
```
        if (_targetToken == NATIVE_TOKEN) {
            if (msg.value < sourceAmount) {
                revert InsufficientNativeTokenSent();
            }
        } else {
            if (msg.value > 0) {
                revert UnnecessaryNativeTokenReceived();
            }
        }
```

Comment (ivanzhelyazkov) [reply contracts/vortex/CarbonVortex.sol:637]: prefer to keep it as it is

**Review by barakman — APPROVED** [review]

**Review by lbeder — APPROVED** [review]

File: contracts/vortex/CarbonVortex.sol [hunk 3]
```diff
@@ -631,6 +631,10 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
         if (sourceAmount == 0) {
             revert InvalidTrade();
         }
+        // revert if unnecessary native token is received
+        if (_targetToken != NATIVE_TOKEN && msg.value > 0) {
```
Comment (lbeder) [inline contracts/vortex/CarbonVortex.sol:635]: You can refactor the `!=` and `==` check into a single flow, but it's not critical
